src/documentation/0000-socket-programming-with-net-module/index.md

```diff
@@ -0,0 +1,134 @@
+
+```js
+const net = require('net')
+
+// Create a TCP server using the `net.createServer()` method
+// The callback method is an event listener for the 'connection' event
+// This method is passed a single parameter, the client socket connection
+const server = net.createServer(client => {
+
+  console.log('Client connected!')
+
+  // Manual implementation of an echo server:
+  // Get data in from the client, and write it back to the client without modifying it.
+  client.on('data', buf => { // receiving data from the client connection as a Buffer
+    console.log(buf.toString()) // easily transfrom this into a string with the Buffer.toString() prototype method
+    client.write(buf)
+  })
+
+  // Automatic implementation of an echo server:
+  client.pipe(client) // client connections are streams so you have full access to regular stream methods
+})
+
+// Start the server on localhost:8124
+// The `http` module, inherits this method for its own http.Server.listen
+server.listen(8124, 'localhost', () => {
+  console.log('Waiting for connection...')
+})
+```
+
+With this server running, connect to it using your system's command-line TCP interface. The two most common are telnet and netcat. If you're unsure which, try both! If neither work a quick search for "telnet alternative for \<your OS\>" will help. 
+
+In another terminal connect to the running TCP server:
+```bash
+telnet localhost 8124
+# or
+nc localhost 8124
+```
+
+If you see the message `Client connected!` in the TCP server terminal, you can begin sending messages from the client connection terminal. Whatever you send should be sent back to you immediatly (if you copy-and-pasted the code from above directly, you'll receive your message twice because their are two echo implementations). Disconnecting is easy; use your terminal's quit keystroke (commonly `CTRL+C`) or exit the terminal application instance.
+
+Great work! You have succesfully implemented your first TCP server-client connection.
+
+## Example using `net.createConnection()`
+
+Before starting, open to terminals and startup the Node.js repl (you can also use two files called `server.js` and `client.js` respectively). In the following example, we will create a TCP server and a connection to that server using the `net` module. By doing so, the interactivity is not the same as the previous echo example. In fact, without using another core module such as `readline` or `fs`, this example is completely non-interactive. The client code will disconnect itself at the end of its execution.
+
+```js
+// In the first repl, create and start a server
+const net = require('net')
+
+const server = net.createServer(client => {
+
+  console.log('Client connected') // log the client has connected to the server 
+
+  client.write('Hello from the server!\n') // say hello to the new connection
+
+  client.on('data', buf => {
+    const str = buf.toString()
```

Comment:
A relatively common issue that I see with Node.js code is this pattern of using `.on('data')` with Buffers and converting them to strings inside the event handler – that is buggy in most cases, because it breaks down when UTF-8 characters end up split across chunks.

I would be wary of recommending something here that works most of the time, but not always. The “proper” solution would be calling `client.setEncoding('utf8');` outside of the handler, which makes the `data` event receive strings rather than buffers, and I think that that is suitable for this guide.

Reply:
I didn't know this! I'll edit the guide to explain this

Reply:
I modified the code to use this and added a comment